python/rmm/_lib/memory_resource.pyx

```diff
@@ -271,43 +323,118 @@ cpdef _initialize(
+        else:
+            mr = typ(*args)
 
-cpdef _set_default_resource(MemoryResource mr):
+        _set_per_device_resource(device, mr)
+    
+    # reset CUDA device to original
+    set_current_device(original_device)
+
+
+cpdef get_per_device_resource(int device):
     """
-    Set the memory resource to use for RMM device allocations.
+    Get the default memory resource for the specified device.
 
     Parameters
     ----------
+    device : int
+        The ID of the device for which to get the memory resource.
+    """
+    global _per_device_mrs
+    return _per_device_mrs[device]
+
+
+cpdef _set_per_device_resource(int device, MemoryResource mr):
+    """
+    Set the default memory resource for the specified device.
+
+    Parameters
+    ----------
+    device : int
+        The ID of the device for which to get the memory resource.
     mr : MemoryResource
-        A MemoryResource object. See `rmm.mr` for the different
-        MemoryResource types available.
+        The memory resource to set.
     """
-    global _mr
-    _mr = mr
-    set_default_resource(_mr.c_obj)
+    global _per_device_mrs
+    _per_device_mrs[device] = mr
+    _mr = mr  # coerce Python object to C object 
+    set_per_device_resource(device, _mr.c_obj)
 
 
-cpdef get_default_resource_type():
+cpdef set_current_device_resource(MemoryResource mr):
     """
-    Get the default memory resource type used for RMM device allocations.
+    Set the default memory resource for the current device.
+
+    Parameters
+    ----------
+    mr : MemoryResource
+        The memory resource to set.
     """
-    return type(_mr)
+    _set_per_device_resource(get_current_device(), mr)
+
+
+cpdef get_per_device_resource_type(int device):
+    """
+    Get the memory resource type used for RMM device allocations on the
+    specified device.
+
+    Parameters
+    ----------
+    device : int
+        The device ID
+    """
+    return type(get_per_device_resource(device))
+
+
+cpdef get_current_device_resource():
+    return get_per_device_resource(get_current_device())
+
+
+cpdef get_current_device_resource_type():
+    """
+    Get the memory resource type used for RMM device allocations on the
+    current device.
+    """
+    return type(get_current_device_resource())
 
 
 cpdef is_initialized():
-    global _mr
-    return _mr.c_obj.get() is not NULL
+    """
+    Check whether RMM is initialized
+    """
+    global _per_device_mrs
+    cdef MemoryResource each_mr
+    return all(
+        [each_mr.c_obj.get() is not NULL
+            for each_mr in _per_device_mrs.values()]
+    )   
 
 
 cpdef _flush_logs():
-    global _mr
-    _mr.flush()
+    """
+    Flush the logs of all currently initialized LoggingResourceAdaptor
+    memory resources
+    """
+    global _per_device_mrs
+    cdef MemoryResource each_mr
+    [each_mr.flush() for each_mr in _per_device_mrs.values()
+        if type(each_mr) is LoggingResourceAdaptor]
```

Comment:
No need to create a list unnecessarily here, also we should check for an instance of the class in case someone creates their own class around the `LoggingResourceAdaptor` class:
```suggestion
    for each_mr in _per_device_mrs.values():
        if isinstance(each_mr, LoggingResourceAdaptor):
            each_mr.flush()
```